extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQCompactionRunner.java

```diff
@@ -0,0 +1,508 @@
+  ) throws Exception
+  {
+    List<MSQControllerTask> msqControllerTasks = compactionToMSQTasks(compactionTask, intervalDataSchemas);
+
+    if (msqControllerTasks.isEmpty()) {
+      String msg = StringUtils.format(
+          "Can't find segments from inputSpec[%s], nothing to do.",
+          compactionTask.getIoConfig().getInputSpec()
+      );
+      return TaskStatus.failure(compactionTask.getId(), msg);
+    }
+    return runSubtasks(
+        msqControllerTasks,
+        taskToolbox,
+        currentSubTaskHolder,
+        compactionTask.getId()
+    );
+  }
+
+  public List<MSQControllerTask> compactionToMSQTasks(
+      CompactionTask compactionTask,
+      List<NonnullPair<Interval, DataSchema>> intervalDataSchemas
+  ) throws JsonProcessingException
+  {
+    List<MSQControllerTask> msqControllerTasks = new ArrayList<>();
+
+    for (NonnullPair<Interval, DataSchema> intervalDataSchema : intervalDataSchemas) {
+      Query<?> query;
+      Interval interval = intervalDataSchema.lhs;
+      DataSchema dataSchema = intervalDataSchema.rhs;
+
+      if (!isGroupBy(dataSchema)) {
+        query = buildScanQuery(compactionTask, interval, dataSchema);
+      } else {
+        query = buildGroupByQuery(compactionTask, interval, dataSchema);
+      }
+      QueryContext compactionTaskContext = new QueryContext(compactionTask.getContext());
+
+      MSQSpec msqSpec = MSQSpec.builder()
+                               .query(query)
+                               .columnMappings(getColumnMappings(dataSchema))
+                               .destination(buildMSQDestination(compactionTask, dataSchema, compactionTaskContext))
+                               .assignmentStrategy(MultiStageQueryContext.getAssignmentStrategy(compactionTaskContext))
+                               .tuningConfig(buildMSQTuningConfig(compactionTask, compactionTaskContext))
+                               .build();
+
+      Map<String, Object> msqControllerTaskContext = createMSQTaskContext(compactionTask, dataSchema);
+
+      MSQControllerTask controllerTask = new MSQControllerTask(
+          compactionTask.getId(),
+          msqSpec.withOverriddenContext(msqControllerTaskContext),
+          null,
+          msqControllerTaskContext,
+          null,
+          null,
+          null,
+          msqControllerTaskContext,
+          injector
+      );
+      msqControllerTasks.add(controllerTask);
+    }
+    return msqControllerTasks;
+  }
+
+  private static DataSourceMSQDestination buildMSQDestination(
+      CompactionTask compactionTask,
+      DataSchema dataSchema,
+      QueryContext compactionTaskContext
+  )
+  {
+    final Interval replaceInterval = compactionTask.getIoConfig()
+                                                   .getInputSpec()
+                                                   .findInterval(compactionTask.getDataSource());
+
+    final List<String> segmentSortOrder = MultiStageQueryContext.getSortOrder(compactionTaskContext);
```

Comment:
What's populating the sort order originally? Should we even be allowing it for auto-compaction, since I think the native auto-compaction doesn't support it.

Reply:
I think that makes sense. Have removed it.